src/lib/precondition.ts

```diff
@@ -0,0 +1,375 @@
+        ];
+      })
+    );
+  } else throw Error('bug');
+}
+
+function preconditionSubclass<
+  K extends LongKey,
+  U extends FlatPreconditionValue[K]
+>(
+  party: Party,
+  longKey: K,
+  fieldType: AsFieldElements<U>,
+  { read, vars, constrained }: PreconditionContext
+) {
+  return {
+    get() {
+      read.add(longKey);
+      return (vars[longKey] ??
+        (vars[longKey] = getVariable(party, longKey, fieldType))) as U;
+    },
+    assertEquals(value: U) {
+      constrained.add(longKey);
+      let property = getPath(
+        party.body.preconditions,
+        longKey
+      ) as AnyCondition<U>;
+      if ('isSome' in property) {
+        property.isSome = Bool(true);
+        property.value = value;
+      } else if ('lower' in property) {
+        property.lower = value;
+        property.upper = value;
+      } else {
+        setPath(party.body.preconditions, longKey, value);
+      }
+    },
+    assertNothing() {
+      constrained.add(longKey);
+    },
+  };
+}
+
+function getVariable<K extends LongKey, U extends FlatPreconditionValue[K]>(
+  party: Party,
+  longKey: K,
+  fieldType: AsFieldElements<U>
+): U {
+  // in compile, just return an empty variable
+  if (GlobalContext.inCompile()) {
+    return Circuit.witness(fieldType, (): U => {
+      throw Error(
+        `This error is thrown because you are reading out the value of a variable, when that value is not known.
+To write a correct circuit, you must avoid any dependency on the concrete value of variables.`
+      );
+    });
+  }
+  // if not in compile, get the variable's value first
+  let [accountOrNetwork, ...rest] = longKey.split('.');
+  let key = rest.join('.');
+  let value: U;
+  if (accountOrNetwork === 'account') {
+    let address = party.body.publicKey;
+    let account: any = Mina.getAccount(address);
+    value = account[key];
+    if (value === undefined)
+      throw Error(
+        `Could not get \`${key}\` on account with public key ${address.toBase58()}. The property may not be available on this account.`
+      );
+  } else if (accountOrNetwork === 'network') {
+    let networkState = Mina.getNetworkState();
+    value = getPath(networkState, key);
+  } else {
+    throw Error('impossible');
+  }
+  // in prover, return a new variable which holds the value
+  // outside, just return the value
+  if (GlobalContext.inProver()) {
+    return Circuit.witness(fieldType, () => value);
+  } else {
+    return value;
+  }
+}
+
+// per-party context for checking invariants on precondition construction
+type PreconditionContext = {
+  isSelf: boolean;
+  vars: Partial<FlatPreconditionValue>;
+  read: Set<LongKey>;
+  constrained: Set<LongKey>;
+};
+
+function initializePreconditions(party: Party, isSelf: boolean) {
+  preconditionContexts.set(party, {
+    read: new Set(),
+    constrained: new Set(),
+    vars: {},
+    isSelf,
+  });
+}
+
+function assertPreconditionInvariants(party: Party) {
+  let context = getPreconditionContextExn(party);
```

Comment:
this is the function that throws an error if the user used `.get()` on a account / network field, without explicitly adding a precondition. calling `get` would have added an entry to `context.read`. adding a precondition would have added an entry to `context.constrained`.